### PR TITLE
Increase export throughput

### PR DESCRIPTION
**Pull request by trask**
The first commit is reverting the previous implementation (which was a mess from a diff perspective), so recommend reviewing just the second commit which contains the new implementation, which is much better from a diff perspective, and aligns with the approach we will probably end up with upstream in otel.

### CODE REVIEW

File: agent/agent-tooling/src/main/java/com/microsoft/applicationinsights/agent/internal/telemetry/BatchSpanProcessor.java [hunk 1]
```diff
@@ -206,8 +216,12 @@ private void flush() {
         }
       }
       exportCurrentBatch();
-      flushRequested.get().succeed();
-      flushRequested.set(null);
+      CompletableResultCode.ofAll(pendingExports).join(exporterTimeoutNanos, TimeUnit.NANOSECONDS);
+      CompletableResultCode flushResult = flushRequested.get();
+      if (flushResult != null) {
```
Comment (heyams) [inline agent/agent-tooling/src/main/java/com/microsoft/applicationinsights/agent/internal/telemetry/BatchSpanProcessor.java:221]: is null check needed?

File: agent/agent-tooling/src/main/java/com/microsoft/applicationinsights/agent/internal/telemetry/BatchSpanProcessor.java [hunk 2]
```diff
@@ -252,7 +266,15 @@ private void exportCurrentBatch() {
...
         CompletableResultCode result = spanExporter.send(Collections.unmodifiableList(batch));
-        result.join(exporterTimeoutNanos, TimeUnit.NANOSECONDS);
+        if (pendingExports.size() < maxPendingExports - 1) {
+          pendingExports.add(result);
+          result.whenComplete(
+              () -> {
+                pendingExports.remove(result);
+              });
+        } else {
+          result.join(exporterTimeoutNanos, TimeUnit.NANOSECONDS);
```
Comment (heyams) [inline agent/agent-tooling/src/main/java/com/microsoft/applicationinsights/agent/internal/telemetry/BatchSpanProcessor.java:276]: can we add log here to warn users that the maxPendingExports is reached here?

File: agent/agent-tooling/src/main/java/com/microsoft/applicationinsights/agent/internal/telemetry/BatchSpanProcessor.java (hunk 1 above)
Comment (trask) [reply agent/agent-tooling/src/main/java/com/microsoft/applicationinsights/agent/internal/telemetry/BatchSpanProcessor.java:221]: I pulled this change in from upstream

Comment (heyams) [reply agent/agent-tooling/src/main/java/com/microsoft/applicationinsights/agent/internal/telemetry/BatchSpanProcessor.java:221]: maybe upstream needs to be updated? based on the old code, null check doesn't seem to be necessary.

File: agent/agent-tooling/src/main/java/com/microsoft/applicationinsights/agent/internal/telemetry/BatchSpanProcessor.java [hunk 3]
```diff
@@ -267,12 +270,17 @@ private void exportCurrentBatch() {
...
         CompletableResultCode result = spanExporter.send(Collections.unmodifiableList(batch));
         if (pendingExports.size() < maxPendingExports - 1) {
+          addAsyncExport.recordSuccess();
           pendingExports.add(result);
           result.whenComplete(
               () -> {
                 pendingExports.remove(result);
               });
         } else {
+          addAsyncExport.recordFailure(
```
[the remark this answers is not in the document]
Comment (trask) [reply agent/agent-tooling/src/main/java/com/microsoft/applicationinsights/agent/internal/telemetry/BatchSpanProcessor.java:280]: I don't expect this to be common, it allows 100 concurrent exports

File: agent/agent-tooling/src/main/java/com/microsoft/applicationinsights/agent/internal/telemetry/BatchSpanProcessor.java (hunk 1 above)
Comment (trask) [reply agent/agent-tooling/src/main/java/com/microsoft/applicationinsights/agent/internal/telemetry/BatchSpanProcessor.java:224]: it looks like upstream added the null check to make for nullaway static analysis happy, I'd rather stay in sync with them

**Review by heyams — APPROVED** [review]

**Pull request merged**